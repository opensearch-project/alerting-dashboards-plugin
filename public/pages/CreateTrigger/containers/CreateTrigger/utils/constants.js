/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const TRIGGER_TYPE = {
  AD: 'anomaly_detector_trigger',
  BUCKET_LEVEL: 'bucket_level_trigger',
  ALERT_TRIGGER: 'alerting_trigger',
  QUERY_LEVEL: 'query_level_trigger',
  DOC_LEVEL: 'document_level_trigger',
};

export const FORMIK_INITIAL_BUCKET_SELECTOR_VALUES = {
  buckets_path: {
    '': '',
  },
  parent_bucket_path: 'composite_agg',
  script: {
    source: 'params.',
  },
};

export const FORMIK_INITIAL_TRIGGER_CONDITION_VALUES = {
  thresholdValue: 10000,
  thresholdEnum: 'ABOVE',
  anomalyDetector: {
    triggerType: TRIGGER_TYPE.AD,
    anomalyGradeThresholdValue: 0.7,
    anomalyGradeThresholdEnum: 'ABOVE',
    anomalyConfidenceThresholdValue: 0.7,
    anomalyConfidenceThresholdEnum: 'ABOVE',
  },
  script: {
    lang: 'painless',
    source: 'ctx.results[0].hits.total.value > 0',
  },
  buckets_path: {},
  parent_bucket_path: 'composite_agg',
  gap_policy: '',
  query: undefined,
  queryMetric: undefined,
  andOrCondition: undefined,
};

export const FORMIK_INITIAL_TRIGGER_VALUES = {
  name: '',
  severity: '1',
  minTimeBetweenExecutions: null,
  rollingWindowSize: null,
  script: {
    lang: 'painless',
    source: `ctx.results[0].hits.total.value > 0`,
  },
  bucketSelector: JSON.stringify(FORMIK_INITIAL_BUCKET_SELECTOR_VALUES, null, 4), // TODO: To be used for Bucket-Level Triggers defined by query
  triggerConditions: [],
  thresholdValue: 10000,
  thresholdEnum: 'ABOVE',
  anomalyDetector: {
    triggerType: TRIGGER_TYPE.AD,
    anomalyGradeThresholdValue: 0.7,
    anomalyGradeThresholdEnum: 'ABOVE',
    anomalyConfidenceThresholdValue: 0.7,
    anomalyConfidenceThresholdEnum: 'ABOVE',
  },
  where: {
    fieldName: [],
    operator: 'includes',
    fieldValue: '',
    fieldRangeStart: 0,
    fieldRangeEnd: 0,
  },
  actions: undefined,
};

export const FORMIK_INITIAL_DOC_LEVEL_SCRIPT = {
  lang: FORMIK_INITIAL_TRIGGER_VALUES.script.lang,
  source: '(query[name=<queryName>] || query[name=<queryName>]) && query[tag=<queryTag>]',
};

export const HITS_TOTAL_RESULTS_PATH = 'ctx.results[0].hits.total.value';
export const AGGREGATION_RESULTS_PATH = 'ctx.results[0].aggregations.metric.value';
export const ANOMALY_GRADE_RESULT_PATH = 'ctx.results[0].aggregations.max_anomaly_grade.value';
export const ANOMALY_CONFIDENCE_RESULT_PATH = 'ctx.results[0].hits.hits[0]._source.confidence';
export const NOT_EMPTY_RESULT =
  'ctx.results != null && ctx.results.length > 0 && ctx.results[0].aggregations != null && ctx.results[0].aggregations.max_anomaly_grade != null && ctx.results[0].hits.total.value > 0 && ctx.results[0].hits.hits[0]._source != null && ctx.results[0].hits.hits[0]._source.confidence != null';
