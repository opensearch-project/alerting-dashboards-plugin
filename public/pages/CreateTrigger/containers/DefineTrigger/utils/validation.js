/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const validateTriggerName = (triggers = [], index, isFlyOut = false) => {
  return (value) => {
    const trimmedValue = value.trim();
    if (!trimmedValue) return isFlyOut ? 'Required.' : 'Trigger name is required.';
    const triggerNameExistWithIndex = triggers.some((trigger, i) => {
      return i !== index && trimmedValue === trigger.name.trim();
    });
    if (triggerNameExistWithIndex) {
      return 'Trigger name already used.';
    }

    // TODO: character restrictions
    // TODO: character limits
  };
};
