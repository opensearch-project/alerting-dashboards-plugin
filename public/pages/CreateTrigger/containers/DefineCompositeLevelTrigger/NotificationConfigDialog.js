/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from 'react';
import {
  EuiSmallButton,
  EuiSpacer,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
} from '@elastic/eui';
import Message from '../../components/Action/actions';
import { FORMIK_INITIAL_ACTION_VALUES } from '../../utils/constants';
import { getTriggerContext } from '../../utils/helper';
import { formikToMonitor } from '../../../CreateMonitor/containers/CreateMonitor/utils/formikToMonitor';
import _ from 'lodash';
import { formikToTrigger } from '../CreateTrigger/utils/formikToTrigger';
import { backendErrorNotification, titleTemplate } from '../../../../utils/helpers';
import { checkForError } from '../ConfigureActions/ConfigureActions';
import { TRIGGER_TYPE } from '../CreateTrigger/utils/constants';
import { getDataSourceId } from '../../../utils/helpers';

const NotificationConfigDialog = ({
  closeModal,
  triggerValues,
  httpClient,
  notifications,
  actionIndex,
  triggerIndex,
  formikFieldPath,
}) => {
  const monitor = formikToMonitor(triggerValues);
  delete monitor.monitor_type;
  const context = getTriggerContext({}, monitor, triggerValues, 0);

  const initialActionValues = _.cloneDeep(FORMIK_INITIAL_ACTION_VALUES);
  let action = _.get(triggerValues, `${formikFieldPath}actions[${actionIndex}]`, {
    ...initialActionValues,
  });

  const [initialValues, setInitialValues] = useState({});

  useEffect(() => {
    setInitialValues({
      [`action${actionIndex}`]: _.get(
        triggerValues,
        `${formikFieldPath}actions.${actionIndex}`,
        ''
      ),
    });
  }, []);

  const sendTestMessage = async (index) => {
    const monitorData = _.cloneDeep(monitor);
    let testTrigger = _.cloneDeep(
      formikToTrigger(triggerValues, monitorData.ui_metadata)[triggerIndex]
    );

    testTrigger = {
      ...testTrigger,
      name: _.get(triggerValues, `${formikFieldPath}name`, ''),
      severity: _.get(triggerValues, `${formikFieldPath}severity`, ''),
    };
    const action = _.get(testTrigger, `${TRIGGER_TYPE.COMPOSITE_LEVEL}.actions[${index}]`);
    const condition = {
      ..._.get(testTrigger, `${TRIGGER_TYPE.COMPOSITE_LEVEL}.condition`),
      script: { lang: 'painless', source: 'return true' },
    };

    delete testTrigger[TRIGGER_TYPE.COMPOSITE_LEVEL];

    _.set(testTrigger, 'actions', [action]);
    _.set(testTrigger, 'condition', condition);

    const testMonitor = { ...monitor, triggers: [{ ...testTrigger }] };

    try {
      const response = await httpClient.post('../api/alerting/monitors/_execute', {
        query: { dryrun: false, dataSourceId: getDataSourceId() },
        body: JSON.stringify(testMonitor),
      });
      let error = null;
      if (response.ok) {
        error = checkForError(response, error);
        if (!_.isEmpty(action.destination_id))
          notifications.toasts.addSuccess(`Test message sent to "${action.name}."`);
      }
      if (error || !response.ok) {
        const errorMessage = error == null ? response.resp : error;
        console.error('There was an error trying to send test message', errorMessage);
        backendErrorNotification(notifications, 'send', 'test message', errorMessage);
      }
    } catch (err) {
      console.error('There was an error trying to send test message', err);
    }
  };

  const clearConfig = () => {
    _.set(
      triggerValues,
      `${formikFieldPath}actions.${actionIndex}`,
      initialValues[`action${actionIndex}`]
    );
    closeModal();
  };

  return (
    <EuiModal onClose={() => clearConfig()}>
      <EuiModalHeader>
        <EuiModalHeaderTitle>
          <h1>Configure notification</h1>
        </EuiModalHeaderTitle>
      </EuiModalHeader>
      <EuiModalBody>
        {titleTemplate('Customize message')}

        <EuiSpacer size={'m'} />

        <Message
          fieldPath={formikFieldPath}
          index={actionIndex}
          values={triggerValues}
          action={action}
          context={{
            ctx: context,
          }}
          sendTestMessage={sendTestMessage}
        />
      </EuiModalBody>
      <EuiModalFooter>
        <EuiSmallButton onClick={() => clearConfig()}>Cancel</EuiSmallButton>
        <EuiSmallButton onClick={() => closeModal()} fill>
          Update
        </EuiSmallButton>
      </EuiModalFooter>
    </EuiModal>
  );
};

export default NotificationConfigDialog;
