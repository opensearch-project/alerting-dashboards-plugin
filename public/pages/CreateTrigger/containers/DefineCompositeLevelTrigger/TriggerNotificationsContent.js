/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Fragment, useState, useEffect } from 'react';
import { EuiPanel, EuiFlexGroup, EuiFlexItem, EuiSmallButtonEmpty, EuiSpacer } from '@elastic/eui';
import { FormikComboBox } from '../../../../components/FormControls';
import NotificationConfigDialog from './NotificationConfigDialog';
import _ from 'lodash';
import { FORMIK_INITIAL_ACTION_VALUES } from '../../utils/constants';
import { NOTIFY_OPTIONS_VALUES } from '../../components/Action/actions/Message';
import { required } from '../../../../utils/validate';
import { getManageChannelsUrl } from '../../../../utils/helpers';

const TriggerNotificationsContent = ({
  action,
  options,
  actionIndex,
  triggerIndex,
  triggerValues,
  httpClient,
  notifications,
  hasNotifications,
  formikFieldPath,
}) => {
  const [selected, setSelected] = useState([]);
  const [isModalVisible, setIsModalVisible] = useState(false);

  useEffect(() => {
    action?.destination_id &&
      setSelected([
        {
          label: action.name,
          value: action.destination_id,
          type: action.config_type,
          description: action.description,
        },
      ]);
  }, [action]);

  const onChange = (selectedOptions) => {
    setSelected(selectedOptions);

    const initialActionValues = _.cloneDeep(FORMIK_INITIAL_ACTION_VALUES);
    _.set(triggerValues, `${formikFieldPath}actions[${actionIndex}]`, {
      ...initialActionValues,
      destination_id: selectedOptions[0]?.value,
      name: selectedOptions[0]?.label,
      subject_template: {
        lang: 'mustache',
        source: 'Monitor {{ctx.monitor.name}} triggered an alert {{ctx.trigger.name}}',
      },
      action_execution_policy: {
        action_execution_scope: NOTIFY_OPTIONS_VALUES.PER_ALERT,
      },
    });
  };

  const onBlur = (e, field, form) => {
    form.setFieldTouched(field.name, true);
    form.setFieldError(field.name, required(form.values[field.name]));
  };

  const showConfig = () => setIsModalVisible(true);

  return (
    <Fragment>
      <EuiPanel hasBorder={false} hasShadow={false}>
        <EuiFlexGroup>
          <EuiFlexItem
            grow={true}
            style={{
              maxWidth: '400px',
            }}
          >
            <FormikComboBox
              name={`channel_name_${triggerIndex}_${actionIndex}`}
              formRow
              fieldProps={{}}
              rowProps={{
                label: 'Notification channel',
                isInvalid: (name, form) => form.touched[name] && !selected?.length,
                error: (name, form) => form.touched[name] && !selected?.length && 'Required.',
              }}
              inputProps={{
                placeholder: 'Select a channel to get notified',
                options: options,
                selectedOptions: selected,
                onChange: (selectedOptions) => onChange(selectedOptions),
                onBlur: (e, field, form) => onBlur(e, field, form),
                singleSelection: { asPlainText: true },
              }}
            />
          </EuiFlexItem>
          <EuiFlexItem grow={false} style={{ paddingLeft: 0, marginLeft: 0 }}>
            <EuiSmallButtonEmpty
              iconType={'popout'}
              style={{ marginTop: '22px' }}
              disabled={!hasNotifications}
              onClick={() => window.open(getManageChannelsUrl())}
            >
              Manage channels
            </EuiSmallButtonEmpty>
          </EuiFlexItem>
        </EuiFlexGroup>
        {selected.length ? (
          <Fragment>
            <EuiSpacer size={'xs'} />
            <EuiSmallButtonEmpty iconType={'pencil'} onClick={() => showConfig()}>
              Configure notification
            </EuiSmallButtonEmpty>
          </Fragment>
        ) : null}
      </EuiPanel>
      {isModalVisible && (
        <NotificationConfigDialog
          closeModal={() => setIsModalVisible(false)}
          triggerValues={triggerValues}
          httpClient={httpClient}
          notifications={notifications}
          actionIndex={actionIndex}
          triggerIndex={triggerIndex}
          formikFieldPath={formikFieldPath}
        />
      )}
    </Fragment>
  );
};

export default TriggerNotificationsContent;
