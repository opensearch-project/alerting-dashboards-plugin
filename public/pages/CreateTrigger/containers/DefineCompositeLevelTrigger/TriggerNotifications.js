/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Fragment, useState, useEffect } from 'react';
import {
  EuiSpacer,
  EuiButton,
  EuiText,
  EuiAccordion,
  EuiHorizontalRule,
  EuiButtonIcon,
  EuiToolTip,
} from '@elastic/eui';
import TriggerNotificationsContent from './TriggerNotificationsContent';
import { MAX_CHANNELS_RESULT_SIZE, OS_NOTIFICATION_PLUGIN } from '../../../../utils/constants';
import { titleTemplate } from '../../../../utils/helpers';

const TriggerNotifications = ({
  httpClient,
  triggerActions = [],
  plugins,
  notifications,
  notificationService,
  triggerValues,
  triggerIndex,
  formikFieldPath,
}) => {
  const [actions, setActions] = useState([]);
  const [options, setOptions] = useState([]);

  useEffect(() => {
    let newActions = [...triggerActions];
    if (_.isEmpty(newActions))
      newActions = [
        {
          name: '',
          id: '',
        },
      ];

    setActions(newActions);

    getChannels().then((channels) => setOptions(channels));
  }, [triggerValues, plugins]);

  const getChannels = async () => {
    const hasNotificationPlugin = plugins.indexOf(OS_NOTIFICATION_PLUGIN) !== -1;

    let channels = [];
    let index = 0;
    const getChannels = async () => {
      const config_types = await notificationService.getServerFeatures();
      const getChannelsQuery = {
        from_index: index,
        max_items: MAX_CHANNELS_RESULT_SIZE,
        config_type: config_types,
        sort_field: 'name',
        sort_order: 'asc',
      };

      const channelsResponse = await notificationService.getChannels(getChannelsQuery);

      channels = channels.concat(
        channelsResponse.items.map((channel) => ({
          label: channel.name,
          value: channel.config_id,
          type: channel.config_type,
          description: channel.description,
        }))
      );

      if (channelsResponse.total && channels.length < channelsResponse.total) {
        index += MAX_CHANNELS_RESULT_SIZE;
        await getChannels();
      }
    };

    if (hasNotificationPlugin) {
      await getChannels();
    }

    return channels;
  };

  const onAddNotification = () => {
    const newActions = [...actions];
    newActions.push({
      label: '',
      value: '',
    });
    setActions(newActions);
  };

  const onRemoveNotification = (idx) => {
    const newActions = [...actions];
    newActions.splice(idx, 1);
    _.set(triggerValues, `${formikFieldPath}actions`, newActions);
    setActions(newActions);
  };

  return (
    <Fragment>
      {titleTemplate('Notifications')}
      <EuiSpacer size={'m'} />
      {actions.length
        ? actions.map((action, actionIndex) => (
            <>
              {actionIndex > 0 && <EuiHorizontalRule margin={'s'} />}
              <EuiAccordion
                title={`Notification ${actionIndex + 1}`}
                key={`notification-accordion-${triggerIndex}-${actionIndex}`}
                id={`notification-accordion-${triggerIndex}-${actionIndex}`}
                initialIsOpen={!actionIndex}
                buttonContent={<EuiText>{`Notification ${actionIndex + 1}`}</EuiText>}
                paddingSize={'s'}
                extraAction={
                  <EuiToolTip content={'Remove notification'}>
                    <EuiButtonIcon
                      color={'danger'}
                      aria-label={'Delete notification'}
                      iconType={'trash'}
                      onClick={() => onRemoveNotification(actionIndex)}
                      size={'s'}
                    />
                  </EuiToolTip>
                }
              >
                <TriggerNotificationsContent
                  action={action}
                  options={options}
                  actionIndex={actionIndex}
                  triggerIndex={triggerIndex}
                  notifications={notifications}
                  triggerValues={triggerValues}
                  httpClient={httpClient}
                  hasNotifications={plugins.indexOf(OS_NOTIFICATION_PLUGIN) !== -1}
                  formikFieldPath={formikFieldPath}
                />
              </EuiAccordion>
            </>
          ))
        : null}
      {actions.length ? <EuiHorizontalRule margin={'s'} /> : null}
      <EuiButton onClick={() => onAddNotification()}>Add notification</EuiButton>
    </Fragment>
  );
};

export default TriggerNotifications;
