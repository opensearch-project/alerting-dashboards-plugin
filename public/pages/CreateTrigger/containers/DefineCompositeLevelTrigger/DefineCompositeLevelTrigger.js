/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { EuiSpacer, EuiTitle, EuiAccordion, EuiButton } from '@elastic/eui';
import { FormikFieldText, FormikSelect } from '../../../../components/FormControls';
import { hasError, isInvalid, required } from '../../../../utils/validate';
import { DEFAULT_TRIGGER_NAME, SEVERITY_OPTIONS } from '../../utils/constants';
import CompositeTriggerCondition from '../../components/CompositeTriggerCondition/CompositeTriggerCondition';
import TriggerNotifications from './TriggerNotifications';
import { validateTriggerName } from '../DefineTrigger/utils/validation';
import { FORMIK_COMPOSITE_INITIAL_TRIGGER_VALUES } from '../CreateTrigger/utils/constants';
import { titleTemplate } from '../../../../utils/helpers';

const defaultRowProps = {
  label: titleTemplate('Trigger name'),
  style: { paddingLeft: '10px' },
  isInvalid,
  error: hasError,
};

const defaultInputProps = { isInvalid };

const selectFieldProps = {
  validate: () => {},
};

const selectRowProps = {
  label: 'Severity level',
  style: { paddingLeft: '10px', marginTop: '0px' },
  isInvalid,
  error: hasError,
};

const selectInputProps = {
  options: SEVERITY_OPTIONS,
};

const propTypes = {
  values: PropTypes.object.isRequired,
  isDarkMode: PropTypes.bool.isRequired,
};

class DefineCompositeLevelTrigger extends Component {
  render() {
    const {
      values,
      httpClient,
      notifications,
      notificationService,
      plugins,
      touched,
      edit,
      triggerIndex,
      triggerArrayHelpers,
    } = this.props;

    const formikFieldPath = `triggerDefinitions[${triggerIndex}].`;
    const triggerName = _.get(values, `${formikFieldPath}name`, 'Trigger');
    const triggerDefinitions = _.get(values, 'triggerDefinitions', []);
    !triggerDefinitions.length &&
      _.set(values, 'triggerDefinitions', [
        {
          ...FORMIK_COMPOSITE_INITIAL_TRIGGER_VALUES,
          ...triggerDefinitions[triggerIndex],
          severity: 1,
          name: triggerName,
        },
      ]);
    const triggerActions = _.get(values, `${formikFieldPath}actions`, []);

    return (
      <EuiAccordion
        id={triggerName}
        buttonContent={
          <EuiTitle size={'s'} data-test-subj={`${formikFieldPath}_triggerAccordion`}>
            <h1>{_.isEmpty(triggerName) ? DEFAULT_TRIGGER_NAME : triggerName}</h1>
          </EuiTitle>
        }
        initialIsOpen={edit ? false : triggerIndex === 0}
        extraAction={
          <EuiButton
            color={'danger'}
            onClick={() => {
              triggerArrayHelpers.remove(triggerIndex);
            }}
            size={'s'}
          >
            Remove trigger
          </EuiButton>
        }
        style={{ paddingBottom: '15px', paddingTop: '10px' }}
      >
        <div style={{ padding: '0px 30px' }}>
          <EuiSpacer size={'m'} />

          <FormikFieldText
            name={`${formikFieldPath}name`}
            fieldProps={{
              validate: (val) => validateTriggerName(values?.triggerDefinitions, triggerIndex)(val),
            }}
            formRow
            rowProps={{
              ...defaultRowProps,
              style: {
                paddingLeft: 0,
              },
            }}
            inputProps={{
              ...defaultInputProps,
              value: triggerName,
              'data-test-subj': 'composite-trigger-name',
            }}
          />

          <EuiSpacer size={'l'} />

          <CompositeTriggerCondition
            triggerIndex={triggerIndex}
            edit={edit}
            formikFieldPath={formikFieldPath}
            formikFieldName={`triggerConditions`}
            label={titleTemplate(
              'Trigger condition',
              'When selected monitors meet the specified conditions for alert generation, the composite monitor triggers its own alert.'
            )}
            values={values}
            touched={touched}
            isDarkMode={this.props.isDarkMode}
            httpClient={httpClient}
          />

          <EuiSpacer size={'l'} />

          {titleTemplate('Alert severity')}
          <FormikSelect
            name={`${formikFieldPath}severity`}
            formRow
            fieldProps={selectFieldProps}
            rowProps={selectRowProps}
            inputProps={selectInputProps}
          />

          <EuiSpacer size={'xl'} />

          <TriggerNotifications
            formikFieldPath={formikFieldPath}
            triggerIndex={triggerIndex}
            httpClient={httpClient}
            plugins={plugins}
            notifications={notifications}
            notificationService={notificationService}
            triggerValues={values}
            triggerActions={triggerActions}
          />
        </div>
      </EuiAccordion>
    );
  }
}

DefineCompositeLevelTrigger.propTypes = propTypes;

export default DefineCompositeLevelTrigger;
