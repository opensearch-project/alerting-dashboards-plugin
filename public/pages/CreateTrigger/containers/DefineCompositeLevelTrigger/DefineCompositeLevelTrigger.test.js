/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from 'enzyme';

import { Formik } from 'formik';
import { FORMIK_INITIAL_VALUES } from '../../../CreateMonitor/containers/CreateMonitor/utils/constants';
import DefineCompositeLevelTrigger from './DefineCompositeLevelTrigger';

describe('DefineCompositeLevelTrigger', () => {
  test('renders', () => {
    const component = (
      <Formik initialValues={FORMIK_INITIAL_VALUES} onSubmit={() => {}}>
        <DefineCompositeLevelTrigger
          isDarkMode={false}
          httpClient={{}}
          notifications={{}}
          notificationService={{}}
          plugins={{}}
          values={{}}
          touched={{}}
        />
      </Formik>
    );
    expect(render(component)).toMatchSnapshot();
  });
});
