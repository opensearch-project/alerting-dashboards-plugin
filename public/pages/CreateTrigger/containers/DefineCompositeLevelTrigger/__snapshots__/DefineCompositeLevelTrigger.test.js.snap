// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefineCompositeLevelTrigger renders 1`] = `
<div
  class="euiAccordion"
  style="padding-bottom:15px;padding-top:10px"
>
  <div
    class="euiAccordion__triggerWrapper"
  >
    <button
      aria-controls="Trigger"
      aria-expanded="false"
      class="euiAccordion__button"
      id="generated-id"
      type="button"
    >
      <span
        class="euiAccordion__iconWrapper"
      >
        <div>
          EuiIconMock
        </div>
      </span>
      <span
        class="euiIEFlexWrapFix"
      >
        <h1
          class="euiTitle euiTitle--small"
          data-test-subj="triggerDefinitions[undefined]._triggerAccordion"
        >
          Trigger
        </h1>
      </span>
    </button>
    <div
      class="euiAccordion__optionalAction"
    >
      <button
        class="euiButton euiButton--danger euiButton--small"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text"
          >
            Remove trigger
          </span>
        </span>
      </button>
    </div>
  </div>
  <div
    aria-labelledby="generated-id"
    class="euiAccordion__childWrapper"
    id="Trigger"
    role="region"
    tabindex="-1"
  >
    <div>
      <div
        class=""
      >
        <div
          style="padding:0px 30px"
        >
          <div
            class="euiSpacer euiSpacer--m"
          />
          <div
            class="euiFormRow euiFormRow--compressed"
            id="triggerDefinitions[undefined].name-form-row-row"
            style="padding-left:0"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                aria-invalid="false"
                class="euiFormLabel euiFormRow__label"
                for="triggerDefinitions[undefined].name-form-row"
              >
                <h4
                  class="euiTitle euiTitle--xsmall"
                >
                  Trigger name
                </h4>
              </label>
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--compressed"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldText euiFieldText--compressed"
                    data-test-subj="composite-trigger-name"
                    name="triggerDefinitions[undefined].name"
                    type="text"
                    value="Trigger"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--l"
          />
          <div
            class="euiFormRow euiFormRow--compressed"
            id="triggerConditionFieldsContainer_undefined-form-row-row"
            style="max-width:inherit"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                class="euiFormLabel euiFormRow__label"
                for="triggerConditionFieldsContainer_undefined-form-row"
              >
                <h4
                  class="euiTitle euiTitle--xsmall"
                >
                  Trigger condition
                </h4>
                <div
                  class="euiText euiText--extraSmall"
                >
                  <div
                    class="euiTextColor euiTextColor--subdued"
                  >
                    <p>
                      When selected monitors meet the specified conditions for alert generation, the composite monitor triggers its own alert.
                    </p>
                  </div>
                </div>
              </label>
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiFlexGroup--responsive"
                data-test-subj="triggerDefinitions[undefined].triggerConditionsundefined"
                id="triggerConditionFieldsContainer_undefined-form-row"
              >
                <div
                  class="euiFlexItem"
                >
                  <div
                    class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
                    id="triggerDefinitions_undefined__triggerConditions_undefined_code-form-row-row"
                  >
                    <div
                      class="euiFormRow__labelWrapper"
                    >
                      <label
                        class="euiFormLabel euiFormRow__label"
                        for="triggerDefinitions_undefined__triggerConditions_undefined_code-form-row"
                      >
                        Trigger condition
                      </label>
                    </div>
                    <div
                      class="euiFormRow__fieldWrapper"
                    >
                      <div
                        class="euiCodeEditorWrapper"
                        data-test-subj="compositeTriggerConditionEditor_undefined"
                        style="width:80%;height:300px"
                      >
                        <button
                          class="euiCodeEditorKeyboardHint"
                          data-test-subj="codeEditorHint"
                          id="generated-id"
                        >
                          <p
                            class="euiText"
                          >
                            Press Enter to start editing.
                          </p>
                          <p
                            class="euiText"
                          >
                            When you're done, press Escape to stop editing.
                          </p>
                        </button>
                        <div
                          id="generated-id"
                          style="width:80%;height:300px"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--l"
          />
          <h4
            class="euiTitle euiTitle--xsmall"
          >
            Alert severity
          </h4>
          <div
            class="euiFormRow euiFormRow--compressed"
            id="triggerDefinitions[undefined].severity-form-row-row"
            style="padding-left:10px;margin-top:0px"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                aria-invalid="false"
                class="euiFormLabel euiFormRow__label"
                for="triggerDefinitions[undefined].severity-form-row"
              >
                Severity level
              </label>
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--compressed"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <select
                    class="euiSelect euiSelect--compressed"
                    id="triggerDefinitions[undefined].severity"
                    name="triggerDefinitions[undefined].severity"
                  >
                    <option
                      badgetext="Highest"
                      color="[object Object]"
                      value="1"
                    >
                      1 (Highest)
                    </option>
                    <option
                      badgetext="High"
                      color="[object Object]"
                      value="2"
                    >
                      2 (High)
                    </option>
                    <option
                      badgetext="Medium"
                      color="[object Object]"
                      value="3"
                    >
                      3 (Medium)
                    </option>
                    <option
                      badgetext="Low"
                      color="[object Object]"
                      value="4"
                    >
                      4 (Low)
                    </option>
                    <option
                      badgetext="Lowest"
                      color="[object Object]"
                      value="5"
                    >
                      5 (Lowest)
                    </option>
                  </select>
                  <div
                    class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                  >
                    <span
                      class="euiFormControlLayoutCustomIcon"
                    >
                      <div>
                        EuiIconMock
                      </div>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--xl"
          />
          <h4
            class="euiTitle euiTitle--xsmall"
          >
            Notifications
          </h4>
          <div
            class="euiSpacer euiSpacer--m"
          />
          <button
            class="euiButton euiButton--primary euiButton--small"
            type="button"
          >
            <span
              class="euiButtonContent euiButton__content"
            >
              <span
                class="euiButton__text"
              >
                Add notification
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
